Add helper script to run kube e2e tests

marun opened the pull request:
With the right configuration, the kube e2e test suite can be run against
an openshift deployment.  This change adds a script that makes it easy.
This is especially useful for verifying that kube plugins
(e.g. openshift-sdn) retain the behavior expected of virgin kube.

ncdc commented:
@pmorie

pmorie commented:
+1000 to this.  @marun Have you gotten clean runs with this?  I expect that certain upstream E2Es won't run without modification.

pmorie commented:
@deads2k fyi

marun commented:
@pmorie I'm only running a few of the networking tests thus far, and most run without modification.  Some tests require > 1 node (e.g. intra-pod communication), and that's likely true of other tests too.  Regardless, I think this helper script is a good starting point for efforts to run all or part of the kube e2e suite as part of origin CI.

pmorie commented:
@marun Absolutely this is a great start.

I've been running the e2e tests from upstream against origin to test patches, but some of the semantics around service accounts and secrets are different enough in upstream that the e2e test for secrets isn't going to work without modification.

Long-term, I would like us to actually Godep the E2E code for upstream and put hooks in the right places to enable us to add some special-casing for origin in the upstream tests.

marun commented:
@pmorie ++ for being able to vendor the tests in the origin tree.  Not only will it be required to customize behavior but it would be undesirable for upstream kube e2e changes to hit our CI without oversight.

pmorie commented:
@marun Just to be clear I'm not proposing we carry patches to the tests.  We need an API to decorate certain actions in the upstream E2Es.  So, we would:

1.  Godep the upstream test code
2.  Have a package in origin for the necessary plugins to the tests
3.  Run the upstream E2Es in the context of our origin test plugins

Does that make sense?

marun commented:
@pmorie Yes, your intention was clear.

pmorie commented:
@marun This script could probably use a big fat disclaimer that not everything from upstream is expected to work without modification.

marun commented:
@pmorie done.

smarterclayton commented on hack/run-kube-e2e.sh:
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
```
Rename to hack/test-kube-e2e.sh

marun replied:
done.

marun commented:
cc: @rajatchopra

rajatchopra commented:
[merge]

marun commented:
The latest patchset limits origin configuration to OS_CONF_ROOT (defaulting to ORIGIN_ROOT) so that all cluster configuration is read from kubeconfig.

danmcp commented:
@rajatchopra Please review

rajatchopra commented:
[test]

danmcp commented:
[merge]

smarterclayton commented:
I don't know that we need this any more - isn't this just test/extended/core.sh -focus="<something for kube>"?

marun commented:
@smarterclayton The script runs the kubernetes tests from a kubernetes tree, not the ones in the origin tree.  That was useful for me before I added support for running the networking e2e tests out of the origin tree, but maybe that isn't so useful now that vendored code can be trivially run?

There may still be value in having the option of targeting an existing cluster, as the script allows.  I found that useful enough for test development and debugging that I added explicit support to the dind script:

https://github.com/marun/origin/blob/master/hack/dind-cluster.sh#L255
